### ISSUE TITLE
Installer - Allow /CloseApplications to do what it needs to ensure that silent install/upgrades complete successfully

### ISSUE THREAD

**Issue by ferventcoder** [git-for-windows/git#312]
As a user installing/upgrading silently, if something like ssh-agent is running, it can cause the process to hang or fail.  It would be nice to be able to pass something like `/CLOSEAPPLICATIONS` to signify that "Yes, I am sure I would like everything closed so the install/upgrade can continue".

For a GUI based installer, I agree with @sschuberth on ["it simply does not feel right to kill ssh-agent without asking the user first."](https://github.com/msysgit/msysgit/issues/249#issuecomment-53799745) However, for a silent install, a user should be able to pass something that signifies that they are okay with moving forward with whatever is necessary to make the install/upgrade happen successfully.

The relevant code blocks look to have moved over to https://github.com/git-for-windows/build-extra/blob/9f0c810c3872e335218833157272cde7d9f6b7a8/installer/install.iss.in#L349-L418

Follow up from https://github.com/msysgit/msysgit/issues/249.

**Comment by dscho** [comment]
Thanks for following up. I agree with the reasoning. @ferventcoder have you managed to build an installer using the new [Git for Windows SDK](http://git-for-windows.github.io/#download-sdk) yet? If so, could you just patch in the relevant `usr\bin\ssh-agent.exe` lines and see whether it Does The Right Thing?

**Comment by ferventcoder** [comment]
I'm good to help out, I haven't attempted to build with the new SDK yet. :)

**Comment by dscho** [comment]
@ferventcoder could you do that? I'll need you to test things with the new SDK anyway...

**Comment by dscho** [comment]
@ferventcoder once you installed [the Git for Windows SDK](http://git-for-windows.github.io/#download-sdk), you can [make an installer](https://github.com/git-for-windows/git/wiki/Making-an-installer) using the [`silent-install` branch](https://github.com/git-for-windows/build-extra/compare/silent-install).

If that manages to find the `ssh-agent` and close it, I will merge said branch. Could you please test?

**Comment by ferventcoder** [comment]
Will do. Hopefully today.

**Comment by ferventcoder** [comment]
Attempted to extract produced an issue:
![image](https://cloud.githubusercontent.com/assets/63502/9468392/f9cb7884-4b06-11e5-9c75-329b17aacbfb.png)

I don't need a proxy.

**Comment by ferventcoder** [comment]
I think a reboot may be necessary. 

`ping dl.bintray.com` times out.

**Comment by dscho** [comment]
> ping dl.bintray.com times out.

That is correct, `dl.bintray.com` does not answer.

But does it resolve, that is the question... If in doubt, you could either switch to Google's DNS (8.8.8.8) or put the entry "5.153.24.114 dl.bintray.com" into your `C:\Windows\system32\drivers\etc\hosts`.

Or maybe just restarting the SDK installer will work around that problem?

**Comment by ferventcoder** [comment]
> Or maybe just restarting the SDK installer will work around that problem?

Gave it a couple of shots. I decided on a reboot to see if that clears things out. The next step is to flush the dns cache, as that was not the IP address that the machine in question attempting to ping. :/

Usually use Google's DNS, but I'll check for sure. I believe it is a local machine issue.

**Comment by ferventcoder** [comment]
Do you think the issue is possibly a permission-related issue with regards to bintray? 

These items don't show up, but that doesn't mean they won't work.
Based on mirrorlist.git-for-windows:
* https://bintray.com/git-for-windows/pacman/mingw/
* https://bintray.com/git-for-windows/pacman/msys2/

These don't have any public direct downloads:
* https://bintray.com/git-for-windows/pacman/package-database/view
* https://bintray.com/git-for-windows/pacman/pacman/view

**Comment by ferventcoder** [comment]
No, that's not the issue. If I keep trying enough times, it works. :/ Definitely something with my machine.

**Comment by ferventcoder** [comment]
Okay, got something built, will test in a little while and give you a heads up. :)

**Comment by dscho** [comment]
Thanks!

**Comment by ferventcoder** [comment]
Doesn't quite look like that worked

![image](https://cloud.githubusercontent.com/assets/63502/9500644/32000d92-4bec-11e5-96fb-aeb216e86163.png)

**Comment by dscho** [comment]
Darn. Will have a look tomorrow.

**Comment by ferventcoder** [comment]
@dscho I wonder if the uninstall was coming from the OLDer git that was installed? Perhaps I should generate two installers and try this. One to install and run, and the next for upgrade?

**Comment by dscho** [comment]
> I wonder if the uninstall was coming from the OLDer git that was installed?

Oh my, you are *absolutely* correct!

> Perhaps I should generate two installers and try this. One to install and run, and the next for upgrade?

Actually, one and the same installer should be enough. You *can* reinstall one and the same version, e.g. to reinitialize messed-up installations or to change the settings in the wizard pages.

So maybe you can install it once (without a running ssh-agent), then start the ssh-agent, and then try the silent installation again?

Thank you so much!

**Comment by ferventcoder** [comment]
Will do.

**Comment by dscho** [comment]
Thank you!

**Comment by TheBigBear** [comment]
@ferventcoder @dscho I am also a users that is installing/upgrading silently, does this mean that there is a ```/silent``` and a ```/silent-uninstall``` option in the works and coming out real soon now? ;-) 

The github support person, Scott Sanicki, that I asked this question via email pointed me to the ```Squirrel.Windows``` project and said that the Github Desktop software was going to switch to that installer to finally get a working ```/silent``` install option?

What cpomes first? This ```/silent``` and ```/uninstall``` under discussion here, or the switch to the ```Squirrel.Windows``` (ClickOnce act-a-like but better) installer?

quote from email from "Scott Sanicki (Github staff)"
```
Unfortunately we do not have a standalone installer at this time. GitHub Desktop makes use of Microsoft's ClickOnce technology for installation and updates.

We are currently working on an open source replacement for ClickOnce here: https://github.com/squirrel/squirrel.windows (We're always looking for contributors to help!)

Once that technology is complete and ready for use, we hope to switch GitHub Desktop to use that. It would allow for a standalone installer.
```

**Comment by dscho** [comment]
@TheBigBear please note that GitHub for Windows and Git for Windows are different projects. It is true that GitHub for Windows bundles a portable version of Git for Windows, but that does not include the installer that *this* ticket is about.

**Comment by TheBigBear** [comment]
@dscho so this one relates to 'git-for-windows' (dooh read the URL!) ;-) aka the friendly fork of msysgit which has just come out as the first windows version of a cmd line ```git``` in that line of succession without a 'preview' in the naming and came out right away with version number ```2.5.0```?

Ah, I see, sorry for the confusion, my bad. I hadn't noticed the issues of the silent install and uninstalls hanging under certain conditions. (Note to self - more testing!)

Thanks @ferventcoder for coming to the resuce (once again), your efforts are really much appreciated.

**Comment by ferventcoder** [comment]
@TheBigBear To answer the question about Squirrel for Windows, it already has a silent switch - `-s`.

**Comment by TheBigBear** [comment]
@ferventcoder yes, thanks, I know and like that part of it. 
[But as an IT admin I don't really like the concept of 'user installs' at all - but that is an entirely differnet topic alltogether. ;-)]

But Squirel.Windows installer is ```not yet``` used as the installer of ```Github Desktop``` for windows. 

ATM it only has the "ugly" "ugly" MS ClickOnce installer which can't be kept quiet or free of user interaction. (ugly not from aestehics or visual point of view, but from angle of IT admin seeking backgound silent mass deployment)

Rob sure you remember choco had to resort to using autohotkey macros - to get it installed quietly, and uninstall in a quiet way is near impssoible as there can be soo many reasons why it doesn't want to do it without interaction.

Sorry for using the 'bandwidth' in this (almost) totally un-related topic, for this long rant.

**Comment by ferventcoder** [comment]
@dscho It's never as easy as we'd hope. :/

`Git-2.5.1-test-64-bit.exe /VERYSILENT /NORESTART /NOCANCEL /SP- /CLOSEAPPLICATIONS /RESTARTAPPLICATIONS /NOICONS /COMPONENTS="icons,icons\quicklaunch,ext,ext\reg,ext\reg\shellhere,ext\reg\guihere,assoc,assoc_sh" /LOG`

![image](https://cloud.githubusercontent.com/assets/63502/9546292/337bd306-4d57-11e5-9105-bfd0e29524bf.png)

![image](https://cloud.githubusercontent.com/assets/63502/9546298/3a39d09e-4d57-11e5-90bd-43c2ac4e1d18.png)

I wonder if InnoSetup has the concept of running arbitrary commands?

Here's what I would see it do, not even sure it is legal yet though:

`taskkill /F /IM ssh-agent.exe /T`

![image](https://cloud.githubusercontent.com/assets/63502/9546377/d0f86068-4d57-11e5-909f-a27a0729e55d.png)

**Comment by ferventcoder** [comment]
Updated taskkill to the full path using systemroot environment variable.

**Comment by dscho** [comment]
> It's never as easy as we'd hope. :/

Bummer!

> Here's what I would see it do, not even sure it is legal yet though:
> 
> taskkill /F /IM ssh-agent.exe /T

We have to be a little careful here because people might have another ssh-agent running, too, and we must not kill any ssh-agent that did not originate from Git for Windows...

**Comment by dscho** [comment]
Oh, wait a minute... Why is your `ssh-agent` running as `vagrant`? Are you sure it is the one from Git for Windows? You should be able to find out by having *another* (portable) Git for Windows somewhere and calling its `ps -W | grep ssh-agent` (which outputs the full path of `ssh-agent`).

**Comment by ferventcoder** [comment]
> I wonder if InnoSetup has the concept of running arbitrary commands?

Looks like it is possible - http://www.jrsoftware.org/ishelp/index.php?topic=isxfunc_exec
You are probably already familiar with that though.

> We have to be a little careful here because people might have another ssh-agent running, too, and we must not kill any ssh-agent that did not originate from Git for Windows...

How do we know the difference? Or is that where the problem lies? 

So two questions -

1. Is killing all ssh-agent processes harmful (even the ones Git doesn't own) or just a minor annoyance?
1. If we kill off all ssh-agent processes, what side effects could occur?

**Comment by dscho** [comment]
> If we kill off all ssh-agent processes, what side effects could occur?

That would interfere with other applications that bring their own ssh-agent... I would *really* like to avoid that (I remember for example when I had to administer an SSH server, long time ago, and carefully started a second SSH daemon so I could shut down the main SSH daemon, and of course the script to shut down the first daemon *killed all SSH daemons*. That was no fun, I can tell you!).

**Comment by dscho** [comment]
So now I am really puzzled... `/CLOSEAPPLICATIONS` kills e.g. `usr/bin/sh.exe`, but not `usr/bin/ssh-agent.exe`?

**Comment by ferventcoder** [comment]
> Why is your ssh-agent running as vagrant?

That's the name of my user on this vagrant box. It looks like the right process:

![image](https://cloud.githubusercontent.com/assets/63502/9546548/2168b4f2-4d59-11e5-8d8c-d4aef585ec09.png)

> `ps -W | grep ssh-agent`

![image](https://cloud.githubusercontent.com/assets/63502/9546585/78808e86-4d59-11e5-9ae4-d40c79a6ffb0.png)

**Comment by dscho** [comment]
Yeah, sorry. Thanks for clarifying!

**Comment by ferventcoder** [comment]
> So now I am really puzzled... /CLOSEAPPLICATIONS kills e.g. usr/bin/sh.exe, but not usr/bin/ssh-agent.exe?

I'm not sure sh was running. I didn't have a bash console running, just regular command line. It's worth giving that a shot though.

**Comment by ferventcoder** [comment]
Trying to stick to tools that are available in stock Windows here to get the ssh-agent processes that we care about for the upgrade.

`wmic process where "name='ssh-agent.exe'" get processid, executablepath`

![image](https://cloud.githubusercontent.com/assets/63502/9546689/6e5cf682-4d5a-11e5-99ea-2df09ae273ee.png)

**Comment by sschuberth** [comment]
I'm only briefly following the conversation, so apologies if this is irrelevant to you, but I wanted to point out that ASAIK the installer is still using the code that I write in [modules.inc.iss](https://github.com/git-for-windows/build-extra/blob/master/installer/modules.inc.iss). It is not yet using Inno Setup's built-in support for the Restart manager API yet. This is because we needed to be able to detect running processes that lock your files before Inno Setup had the feature. That said, the `/CLOSEAPPLICATIONS` switch refers to Inno Setup's build-in support, which might interfere with the logic in `modules.inc.iss`.

**Comment by ferventcoder** [comment]
Removing the silent option, it appears that bash interferes as well.

That is probably expected though.
![image](https://cloud.githubusercontent.com/assets/63502/9546769/0ecd52e2-4d5b-11e5-88c7-88ea284679b5.png)

**Comment by ferventcoder** [comment]
@sschuberth I do remember a conversation about that. :)

**Comment by dscho** [comment]
@sschuberth thanks for noticing! From [this comment](https://github.com/git-for-windows/build-extra/blob/4e13a83d1a386217c021eb1360b063103d052b0c/installer/install.iss.in#L360) I assumed that we use the Restart Manager, though I see that we're using InnoSetup 5.5.6 (and its major version is below 6)...

**Comment by dscho** [comment]
Whoops. The version in question [is actually the Windows version](https://github.com/git-for-windows/build-extra/blob/4e13a83d1a386217c021eb1360b063103d052b0c/installer/install.iss.in#L351), so nevermind!

**Comment by sschuberth** [comment]
> I believe that is why the choco package currently doesn't have `/CLOSEAPPLICATIONS /RESTARTAPPLICATIONS` in it.

Exactly.

To further clarify, my `modules.inc.iss` also uses the Restart Manager API if the Windows version supports it, and uses fallbacks otherwise. It's just that is uses the Restart Manager API directly, not via Inno Setup's now built-in support. Inno Setup has no fallback if the Windows version does not support the Restart Manager API.

**Comment by dscho** [comment]
So my guess is that for some reason, `ssh-agent.exe` might be marked as not restartable [here](https://github.com/git-for-windows/build-extra/blob/4e13a83d1a386217c021eb1360b063103d052b0c/installer/modules.inc.iss#L544).

**Comment by dscho** [comment]
Point in favor of my theory: https://github.com/git-for-windows/build-extra/blob/4e13a83d1a386217c021eb1360b063103d052b0c/installer/install.iss.in#L1067-L1076 (this is responsible for [this message box](https://github.com/git-for-windows/git/issues/312#issuecomment-135764880).

**Comment by sschuberth** [comment]
`ssh-agent.exe` for sure is not restartable, as restartable applications need to link against the Restart Manager API, and I'm very certain that there is no such Windows-specific code in `ssh-agent.exe`.

**Comment by dscho** [comment]
Okay, so we would need to kill it manually without restarting it (it is not restartable for sure because the `SSH_AUTH_SOCK` and `SSH_AGENT_PID` settings cannot be the same after restarting).

**Comment by ferventcoder** [comment]
@dscho @sschuberth What if it were just `/CLOSEAPPLICATIONS` and no restart?

**Comment by sschuberth** [comment]
According to [this comment](https://github.com/jrsoftware/issrc/issues/125#issuecomment-60895401) by the maintainer of Inno Setup that should work.

**Comment by dscho** [comment]
@ferventcoder I fear that `/CLOSEAPPLICATIONS` still would only try to use the Restart Manager (hence you see that `ssh-agent.exe` is not closed).

So what I have in mind is to test whether `CLOSEAPPLICATIONS` is in effect and call `TerminateProcess()` on the `ssh-agent.exe` process if it is in the list and marked not restartable. (Probably should also do that if `RESTARTAPPLICATIONS` is in effect, we cannot restart it, but we can force it to exit).

**Comment by sschuberth** [comment]
> I fear that `/CLOSEAPPLICATIONS` still would only try to use the Restart Manager

No, that should not be the case, see my comment above. Inno Setup is passing `0` (and not `RmShutdownOnlyRegistered`) to [RmShutdown](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373667%28v=vs.85%29.aspx).

**Comment by dscho** [comment]
Where do we use Restart Manager explicitly?

**Comment by sschuberth** [comment]
[Here](https://github.com/git-for-windows/build-extra/blob/master/installer/modules.inc.iss#L426-L563).

**Comment by dscho** [comment]
Actually, we use it [here](https://github.com/git-for-windows/build-extra/blob/4e13a83d1a386217c021eb1360b063103d052b0c/installer/install.iss.in#L1236).

**Comment by dscho** [comment]
So this should properly terminate `ssh-agent` on Vista and later:

```diff
diff --git a/installer/install.iss.in b/installer/install.iss.in
index 5d76d0d..d34c0f4 100644
--- a/installer/install.iss.in
+++ b/installer/install.iss.in
@@ -1233,7 +1233,7 @@ begin
     if CurStep=ssInstall then begin
         // Shutdown locking processes just before the actual installation starts.
         if SessionHandle>0 then try
-            RmShutdown(SessionHandle,RmShutdownOnlyRegistered,0);
+            RmShutdown(SessionHandle,0,0);
         except
             Log('Line {#__LINE__}: RmShutdown not supported.');
         end;
```

@ferventcoder could you give that a try?

**Comment by dscho** [comment]
BTW @sschuberth your comment on https://msdn.microsoft.com/en-us/library/windows/desktop/aa373667%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396 has been very helpful.

**Comment by ferventcoder** [comment]
@dscho Apologies, I will try to give this a shot tomorrow.

**Comment by ferventcoder** [comment]
@dscho no dice. :(

I removed `/RESTARTAPPLICATIONS` and went for only `/CLOSEAPPLICATIONS`

**Comment by ferventcoder** [comment]
With `/VERYSILENT` I get:
![image](https://cloud.githubusercontent.com/assets/63502/9631749/a921259e-5148-11e5-8e8c-24192793cab3.png)


New version installed with the patch (I removed `/VERYSILENT` for now to see what it was selecting):
~~~
PS C:\Users\Administrator> ssh-agent
PS C:\Users\Administrator> git --version
git version 2.5.0.windows.1.2.g5e906e9
PS C:\Users\Administrator> c:\code\temp\Git-2.5.1-test-64-bit.exe /NORESTART /NOCANCEL /SP- /CLOSEAPPLICATIONS /NOICONS
/COMPONENTS="icons,icons\quicklaunch,ext,ext\reg,ext\reg\shellhere,ext\reg\guihere,assoc,assoc_sh" /LOG
~~~

![image](https://cloud.githubusercontent.com/assets/63502/9631715/7914ec46-5148-11e5-947a-9c55f6c1eb54.png)

**Comment by dscho** [comment]
@ferventcoder did you have a chance to test this with the [patch I posted earlier](https://github.com/git-for-windows/git/issues/312#issuecomment-135785529)?

**Comment by ferventcoder** [comment]
That's what the comments after it were in reference to. Starting with [No dice](https://github.com/git-for-windows/git/issues/312#issuecomment-137064372).

**Comment by dscho** [comment]
Okay, thanks for the clarification. I will have to set aside time to give this a hand myself. Thanks!

**Comment by erichexter** [comment]
I found this little nugget in the installer path.   Trying to install this under DSC. in this case there is no UI thread available.


2015-09-23 18:04:19.398   -- File entry --
2015-09-23 18:04:19.398   Dest filename: C:\Program Files\Git\ReleaseNotes.html
2015-09-23 18:04:19.398   Time stamp of our file: 2015-09-18 17:24:34.000
2015-09-23 18:04:19.398   Installing the file.
2015-09-23 18:04:19.398   Successfully installed the file.
2015-09-23 18:04:20.542   Installation process succeeded.
2015-09-23 18:04:20.760   CurStepChanged raised an exception.
2015-09-23 18:04:20.760   Exception message:
2015-09-23 18:04:20.760   Message box (OK):
                          Runtime Error (at 175:8281):
                          
                          IPersistFile::Save failed; code 0x80070003.
                          The system cannot find the path specified.

**Comment by dscho** [comment]
What "nugget"? What's "DSC"? Do you think I can make sense of that log without additional information?

**Comment by ferventcoder** [comment]
nugget == piece of knowledge
DSC == Microsoft PowerShell Desired State Configuration. A configuration management tool from Microsoft that uses the Managed Object Format (MOF) and Windows Management Instrumentation (WMI) to make and ensure changes on a computer.

**Comment by dscho** [comment]
@ferventcoder thank you for the explanation. I still have no clue, though, what that log is supposed to tell me... :innocent:

**Comment by ferventcoder** [comment]
@dscho No worries. I'm not yet sure it applies to this issue. It appears to be something new. It could be that InnoSetup is attempting to use a user interface to do something during the install. And there is no user interface available at all for DSC.

**Comment by dscho** [comment]
If it is something new, i.e. different, it should be

1. described much better, in accordance to [our guidelines](https://github.com/git-for-windows/git/wiki/Issue-reporting-guidelines)
2. a new ticket.

**Comment by erichexter** [comment]
It may be a new use cause or error, but ultimately I guess a variation in
which the installer is not suppressing user interface dialog when the
/CloseAppplications flag is specified.

Eric Hexter

blog | http://Hex.LosTechies.com
info | http://www.linkedin.com/in/erichexter

**Comment by dscho** [comment]
@erichexter it is unclear. The description is unclear (i.e. non-existent). Please put a little effort into the issue, and certainly open a new ticket when already two people who are respected in this project suggested to do so.

**Comment by erichexter** [comment]
will do

Eric Hexter

blog | http://Hex.LosTechies.com
info | http://www.linkedin.com/in/erichexter

**Comment by ferventcoder** [comment]
@erichexter I know you mentioned this in the choco gitter room and I pointed you to this issue as a possible reason that DSC was having issues (due to ssh-agent likely still running). I would try to see if it is this specific case that is causing it by removing Chocolatey and as much of the DSC bits as possible from the equation. Attempt to install the package directly with a DSC process resource and see if it still fails with just `/VERYSILENT`.

**Comment by DarwinJS** [comment]
I am getting this same error when trying to install git via the chocolatey via chef-client running as a service.  Couldn't see it listed as an issue - could someone link it if so?

**Comment by ferventcoder** [comment]
@DarwinJS you can try using the portable package for now.

**Comment by DarwinJS** [comment]
Rib,
Thanks for the suggestion!  This is for dev boxes and CI servers - so I'd
be concerned something at some time will hiccup if it isn't installed with
the standard installer.

D.

**Comment by DarwinJS** [comment]
The error says a folder does not exist, I wonder if the chocolatey package could check for and create that folder before starting the installer.

I don't like workarounds any more than anyone else, but I've found people complaining of this same error back to 1.x versions over a year ago.

**Comment by dscho** [comment]
@DarwinJS in my experience, complaining about an error has never really helped resolve the issue. I see way too much talking and too little doing about this, unfortunately also from your side.

**Comment by DarwinJS** [comment]
dscho - not really an appropriate comment for someone you've never interacted with before.

The very comment you are responding to is offering a possible work around solution and I do not have a complaining tone.

After I sent my last message to Rob, I considered that I could contribute to the effort by simply running procmon and discovering the folder the installer thinks should exist.  There are certain user profile folders that the system user does not have and it has TWO distinct profile folders depending on process bitness "%windir%\system32\config\systemprofile" & "%windir%\syswow64\systemprofile" (chef client service runs under 32-bit ruby.exe)

I could possibly even update the package to precreate the folder with the work around I proposed.

**Comment by dscho** [comment]
@ferventcoder sorry for exploding, but I am quite frustrated, and I hope you understand: I know exactly how to fix this issue, I am just really, really short on time for this project, and it seems as if this ticket is just simply not going anywhere before I get back on a more regular schedule again. Maybe I can be excused for being triggered by yet another comment I feel distracts from the real solution, i.e. make the darned installer simply detect the processes that need killing and Just Do It?

**Comment by DarwinJS** [comment]
I have a couple comments (hopefully helpful).

I looked over the thread in more detail.

[1] I know this is not Windows Installer, but if in Windows Installer you use settings to disable RestartManager, you get the old "FilesInUse" API support.  This means if files are locked, they are queued for rename on reboot and a reboot required exit code is given.

I am thinking that with an automated package manager I would not want it simply killing processes - restart manager itself was designed to service the new install vision that "installers should not consider what they are doing more important than uptime or other runtime activities the machine may be doing."  Personally I feel it is better to just exit with a reboot required exit code - then the sys admin figures out whether and when to take in process kills / system restarts.  However, if a hard kill is desired, I would advise it would require an opt-in switch - not be the default for the setup.exe.  If combined with not using RestartManager - then the sys admin could choose their own poison - kill processes or risk a required reboot.

[2] I think I may be struggling with a distinct, separate problem as it happens on a fresh box that is being built for the first time - a problem with system account context and a folder that is expected by the installer because my message is that a folder does not exist.  Does this make sense or am I mis-interpreting my error?

When I run a trace on this I will feed it back on this thread - if there are two distinct issues, maybe they can both be fixed in the EXE installer and then those benefits will automatically be had by the Chocolatey package as well - does that make sense?

**Comment by dscho** [comment]
> I would not want it simply killing processes

True. The first step, in my mind, however, is to *detect* the processes that block the installation and error out appropriately and early. That is the direction in which I wanted things to go.

> a problem with system account context and a folder that is expected by the installer because my message is that a folder does not exist

If this happens with Git for Windows' unmodified installer, too, please report it as a separate issue. I would really appreciate it in particular a [Minimal, Complete & Verifiable Example](http://stackoverflow.com/help/mcve) how to reproduce the issue, but in general I recommend to follow the [issue reporting guide lines](https://github.com/git-for-windows/git/wiki/Issue-reporting-guidelines) to avoid unnecessary frustration on both sides (i.e. yours and mine). For example, if the issue is the same as reported in #523, it is good to avoid opening a duplicate ticket.

Please understand that these guide lines, as well as my comments, are not designed to annoy anybody, but they are rather designed to save time (because I am *really* busy with something else, so the time I can spend on Git for Windows is *really* limited for now).

**Comment by DarwinJS** [comment]
>> True. The first step, in my mind, however, is to detect the processes that block the installation and error out appropriately and early. That is the direction in which I wanted things to go.

=> This should be built into the windows installation framework you choose (MSI, Squirrel, NSIS) - it is a basic service provision of most of them.  I would look for how to trigger this support.  
If an installer framework did not have this it would be a significant indicator that it is immature.  I know you aren't wanting to call that part of your toolset into reconsideration - but just wanted to let you know there should be zero effort (like a couple settings) to enact this support *within* the installation framework.

a problem with system account context and a folder that is expected by the installer because my message is that a folder does not exist

> I have finished my testing, found the problem and will submit it with the links you provided.

**Comment by ferventcoder** [comment]
@DarwinJS I think that last blockquote is actually a new comment?

**Comment by DarwinJS** [comment]
Yes sorry - fixed.

After searching existing issues for "quicklaunch" I submitted this one: https://github.com/git-for-windows/git/issues/570

**Comment by dscho** [comment]
> I know you aren't wanting to call that part of your toolset into reconsideration

That is incorrect. I am happy to call it into consideration but not without *substantial* assistance from the people who want to see us move away from InnoSetup.

**Comment by DarwinJS** [comment]
Guys - I reviewed the inno setup source at https://github.com/git-for-windows/build-extra/blob/master/installer/install.iss

I do not have the time, nor InnoSetup expertise to make the suggested changes and do tests - but I thought I'd offer the engineering suggestions since they can be difficult to discern with Windows Setup stuff.  So I offer this information for whomever currently works with the InnoSetup for whatever time in the future there might be to look into it.  I am not expecting that the git team has any bandwidth to address this anytime soon.

It seems the install is 99% "per-machine".  The binaries and all *other* shortcuts go to machine locations - not user profiles. (e.g. Start Menu, Desktop).  However "quicklaunch" folder is per-user - because windows does not have a per-machine quicklaunch.

I also noticed at  https://github.com/git-for-windows/build-extra/blob/master/installer/install.iss#L1512 that there is logic to detect if the user has admin rights and put registry keys in HKLM.

To make the install more resilient for per-machine installs I would suggest:
[1] that a similar if statement to the one at https://github.com/git-for-windows/build-extra/blob/master/installer/install.iss#L1512 be used to negate the "icons\quicklaunch" feature.
[2] Just don't support quicklaunch in the installer.

It would also make sense to do the same for logic *anything else* that goes to a user profile - either drop it or redirect it to a machine location (and to log something like "This install is switching to per-machine mode.")

These changes would make the base setup.exe installer more compatible with these scenarios:
1. Push git to developer machines using background automation. (e.g. SCCM deployment) => Git Installed by SYSTEM user, Used by regular user.
2. Push git to CI server using background automation (e.g CHEF client service) => Git Installed by SYSTEM user, (potentially) Used by SYSTEM user .

FYI - as much of a pain as it is - Windows Installer has built-in facilities to handle all these strange edge cases in the same installer.  That is why an MSI per-machine install does a quick "self-heal" (fault-in) of any per-user configuration that is missing for the application, right when the user tries to access the application.  The other installation frameworks have to bludgeon their way through specific use cases to eventually account for these in their architecture - and most don't get all the way there.

**Comment by ferventcoder** [comment]
As compared to NSIS, InnoSetup is one of the most comprehensive installers available for Windows, second only to Windows Installer (MSI). However, InnoSetup is quite a bit friendlier than Windows Installer and by proxy WiX.

**Comment by DarwinJS** [comment]
Johannes,
I am making adjustments to the chocolatey package so that it will drop the 'icons\quicklaunch' option when running under system account and so that it will not hang on message dialogs at anytime (using '/suppressmessageboxes' on installer).

Information used to update the chocolatey package and for the below recommendations comes from the link at the end of this post.

## Suggestion 1
I wanted to suggest that you follow the instructions in the below article and mark all your files in the git inno setup [files] section as "restartreplace".

I believe this will clean up many of your problems when applications are holding files and either: [a] users do not know what to close or [b] the installation is running automated in the background with no ability to have an interactive user respond to requests to close applications.

Inno setup should then generate a "reboot required" exit code.  Most automation engineers know that a "reboot required" exit code is a possibility and should take responsibility to automate for it.  Interactive installs should display a message that a reboot is required.

In the chocolatey package we will use '/suppressmessageboxes' to prevent eternal hangs while running headless - but that fails the install.  If you made the above change, the install would succeed with a reboot required.

## Suggestion 2
I would also like to suggest that you simply eliminate the "icons\quicklaunch' feature since
- it is an above and beyond option not offered by many installers
- technically quicklaunch support was depreciated by Microsoft a while back
- it seems to be one of the only portions the install that assumes a regular user login during install

If you feel quicklaunch should be retained, a secondary helpful position would be to drop the feature from the list if you detect system account.  (An easy check is if the USERNAME environment variable matches the computer name with a literal dollar sign PREpended, then you are running under SYSTEM.)

Happy to answer any questions about these recommendations!
D.

[http://stackoverflow.com/questions/26480722/inno-setup-how-can-i-suppress-files-in-use-dialog-when-install-silently]

**Comment by dscho** [comment]
> I wanted to suggest that you follow the instructions in the below article and mark all your files in the git inno setup [files] section as "restartreplace".

According to http://www.jrsoftware.org/ishelp/index.php?topic=filessection, this does not work when non-admin users install Git for Windows. Or does your reading of that document disagree?

Further, I doubt that you are *really* suggesting to mark *every* file as "restartreplace". This flag would make no sense for any file without `.dll` nor `.exe` suffix.

Lastly, a patch would be the best place to start a discussion about this.

> If you feel quicklaunch should be retained, a secondary helpful position would be to drop the feature from the list if you detect system account. (An easy check is if the USERNAME environment variable matches the computer name with a literal dollar sign PREpended, then you are running under SYSTEM.)

Of course the correct way to implement this should be (note: untested):

```diff
diff --git a/installer/install.iss b/installer/install.iss
index be10f35..9a6b948 100644
--- a/installer/install.iss
+++ b/installer/install.iss
@@ -83,7 +83,7 @@ Name: default; Description: Default installation; Flags: iscustom
 
 [Components]
 Name: icons; Description: Additional icons
-Name: icons\quicklaunch; Description: In the Quick Launch
+Name: icons\quicklaunch; Description: In the Quick Launch; Check: not IsAdminLoggedOn
 Name: icons\desktop; Description: On the Desktop
 Name: ext; Description: Windows Explorer integration; Types: default
 Name: ext\shellhere; Description: Git Bash Here; Types: default
```

**Comment by DarwinJS** [comment]
According to http://www.jrsoftware.org/ishelp/index.php?topic=filessection, this does not work when non-admin users install Git for Windows. Or does your reading of that document disagree?

> no I don't disagree - but you are putting your files in Program Files - you'll always require admin to do that anyway.

Further, I doubt that you are really suggesting to mark every file as "restartreplace". This flag would make no sense for any file without .dll nor .exe suffix.

> marking the files in this way is a contingency approach - the installer will only use it **for the files that are actually locked during any given run of the setup.exe on a specific system**.  So while DLLs and EXEs are most likely to be locked - why take a chance that the user has a config file open or some other unexpected locking behavior is happening when the application is open or when an "integrated" app that leverages git is doing some weird-not-supposed-to-do-that locking of it's files or config files?  You don't have any downside to marking all the files and only upside if unexpected things are happening at that moment on that target system.  A contingency is basically a "plan for the unexpected".

**Comment by dscho** [comment]
> you are putting your files in Program Files

Nope. We do that only if an admin or power user runs the installer.

> So while DLLs and EXEs are most likely to be locked - why take a chance that the user has a config file open

Okay, let's think this through. A user reinstalls Git for Windows. Due to an oversight, a config file is still open in an editor. The installer finishes the upgrade, marking the config file for replacing upon next restart. Now the user goes on to *use* Git. With the old, possibly incorrect config file. Now, I see a problem there.

Sure, the same problem exists if you look at the `.exe` and `.dll` files, so I am growing more and more dubious whether this is actually a good change.

**Comment by DarwinJS** [comment]
> Nope. We do that only if an admin or power user runs the installer.
Ok - even so, you do not gain anything by not taking a contingency against locked files.  It is not more minimal to no support locked file overwrites - especially if you consider the amount of support inquiries it generates.

> Okay, let's think this through. A user reinstalls Git for Windows. Due to an oversight, a config file is still open in an editor. The installer finishes the upgrade, marking the config file for replacing upon next restart. Now the user goes on to use Git. With the old, possibly incorrect config file. Now, I see a problem there.

> Sure, the same problem exists if you look at the .exe and .dll files, so I am growing more and more dubious whether this is actually a good change.

Most installation frameworks do not treat config files as a simple file.  They treat them as a data store and do not apply the same overwrite if locked rules.  If you have actually experienced what you cited (entire config file overwrite losing changes), then you may mistakenly be putting down your initial config file using a file install (inno's "files section") and then using the mechanism for updating it (Inno's INI section http://www.jrsoftware.org/ishelp/topic_inisection.htm) to make a few slight changes.  If you use the config file update mechanism to create *all* the entries in the entire INI file, the installation framework will always edit the file because you have not given it a file in your file specifications.  This is how most installation frameworks deal with INI files and some may be instrumented this way for XML as well.  To treat config files as a simple file copy essentially tells the installation framework that you want it dealt with as a file, not a config store that it must treat as a config namespace where the installer should distinguish the ability to to update individual items when necessary.

My advice is based on more than 15 years of working on windows installation technology and deployment automation and building and teaching courses on it.  I teach these courses to enterprise admins who have to continually fixup "creative" installers that did things their own way rather than how the platform specified them. The things I am citing are used by massive companies to install software on most machines on the planet (like Adobe reader using MSI) and they have high success rates.  

Obviously this specific advice may not apply to your situation as I cannot be fully aware of all the entire history of your journey with the installer - I only offer it to try to make your installer experiences easier.

**Comment by dscho** [comment]
> If you have actually experienced what you cited

No, I did not "cite" it, and I never experienced it. As a maintainer, I am interested in preventing such things from happening altogether.

I am not convinced that changing the installer to try to edit the config files makes sense: we want to overwrite them, after all (they are owned by the Git for Windows package). Besides, editing them would not address the concern of locked files.

In short, I am getting more and more convinced that this `restartreplace` would introduce more problems than it solves.

**Comment by dscho** [comment]
As it happens, I had to implement this feature recently for different reasons. Therefore the originally requested feature request is incidentally addressed.

**Comment by dscho** [comment]
> I wanted to suggest that you follow the instructions in the below article and mark all your files in the git inno setup [files] section as "restartreplace".

Sorry to be coming back after *such* a long time, but I wanted to let you know that we do this, as of git-for-windows/build-extra@09c11c2906c658061aaa8c74a5bdf481a9b30cbf

> I would also like to suggest that you simply eliminate the "icons\quicklaunch' feature

I reconsidered my stance, and would happily accept any Pull Request to make it so.
